wal/src/message_queue_impl/region_meta.rs

```diff
@@ -0,0 +1,595 @@
+            }
+        }
+
+        // Double check is not needed, due to just one task will write the specific
+        // table.
+        let mut inner = self.inner.write().await;
+        debug_assert!(inner
+            .table_metas
+            .insert(table_id, TableMeta::new(table_id))
+            .is_none());
+        // New table, so returned next sequence num is zero.
+        Ok(SequenceNumber::MIN)
+    }
+
+    pub async fn update_after_table_write(
+        &self,
+        table_id: TableId,
+        write_offset_range: (Offset, Offset),
+        updated_num: u64,
+    ) -> Result<()> {
+        let inner = self.inner.read().await;
+        let table_meta = inner
+            .table_metas
+            .get(&table_id)
+            .with_context(|| UpdateAfterWrite {
+                table_id,
+                msg: format!(
+                    "table:{}'s meta not found while update after its write",
+                    table_id
+                ),
+            })?;
+        table_meta
+            .update_after_write(write_offset_range, updated_num)
+            .await;
+
+        inner
+            .local_latest_offset
+            .fetch_add(updated_num as i64, Ordering::Relaxed);
+
+        Ok(())
+    }
+
+    pub async fn mark_table_deleted(
+        &self,
+        table_id: TableId,
+        next_sequence_num: SequenceNumber,
+    ) -> Result<()> {
+        let inner = self.inner.read().await;
+        let table_meta = inner
+            .table_metas
+            .get(&table_id)
+            .with_context(|| MarkDeleted {
+                table_id,
+                msg: format!("table:{}'s meta not found while mark its deleted", table_id),
+            })?;
+        table_meta.mark_deleted(next_sequence_num).await;
+
+        Ok(())
+    }
+
+    /// Scan the table meta entry in it and get the safe(minimum) offset among
+    /// them to return.
+    ///
+    /// NOTICE: Need to freeze the whole region meta on high-level before
+    /// calling.
+    pub async fn get_safe_delete_offset(&self) -> Result<Offset> {
+        let inner = self.inner.read().await;
+        let mut safe_delete_offset = Offset::MAX;
+        let mut high_watermark = 0;
+        // Calc the min offset in message queue.
+        for table_meta in inner.table_metas.values() {
+            let meta_data = table_meta.get_meta_data().await?;
+            if let Some(offset) = meta_data.safe_delete_offset {
+                safe_delete_offset = cmp::min(safe_delete_offset, offset);
+            }
+            high_watermark = cmp::max(high_watermark, meta_data.current_high_watermark);
+        }
+
+        if safe_delete_offset == Offset::MAX {
+            // All tables are in such states:
+            // + has init, but not written
+            // + has written, but not flushed
+            // + has flushed, but not written again
+            // So, we can directly delete it up to the high_watermark.
+            Ok(high_watermark)
```

Comment:
Can't understand why it is safe to delete up to high_watermark if data of tables has written but not flushed.

Reply:
The comment has problems indeed...
The high watermark should be returned just while all tables are flushed but not written again and just initialized.

I have removed this function, it should be the logic in cleaner and can be performed using the data in snapshot.